content/assignments/memory-allocator/tests/ref/test-malloc-block-reuse.ref

```diff
@@ -0,0 +1,81 @@
+os_malloc (['633'])                                                                       = HeapStart + 0x20510
+os_free (['HeapStart + 0x207a8'])                                                         = <void>
+os_malloc (['1000'])                                                                      = HeapStart + 0x207a8
+os_free (['HeapStart + 0x20ba8'])                                                         = <void>
+os_malloc (['2024'])                                                                      = HeapStart + 0x20ba8
+os_free (['HeapStart + 0x213a8'])                                                         = <void>
+os_malloc (['4000'])                                                                      = HeapStart + 0x213a8
+os_free (['HeapStart + 0x213a8'])                                                         = <void>
+os_malloc (['3970'])                                                                      = HeapStart + 0x213a8
+os_free (['HeapStart + 0x20ba8'])                                                         = <void>
+os_malloc (['1994'])                                                                      = HeapStart + 0x20ba8
+os_free (['HeapStart + 0x207a8'])                                                         = <void>
+os_malloc (['970'])                                                                       = HeapStart + 0x207a8
+os_free (['HeapStart + 0x20510'])                                                         = <void>
+os_malloc (['603'])                                                                       = HeapStart + 0x20510
+os_free (['HeapStart + 0x20350'])                                                         = <void>
+os_malloc (['391'])                                                                       = HeapStart + 0x20350
+os_free (['HeapStart + 0x201d8'])                                                         = <void>
+os_malloc (['320'])                                                                       = HeapStart + 0x201d8
+os_free (['HeapStart + 0x20120'])                                                         = <void>
+os_malloc (['130'])                                                                       = HeapStart + 0x20120
+os_free (['HeapStart + 0x200b8'])                                                         = <void>
+os_malloc (['50'])                                                                        = HeapStart + 0x200b8
+os_free (['HeapStart + 0x20078'])                                                         = <void>
+os_malloc (['10'])                                                                        = HeapStart + 0x20078
+os_free (['HeapStart + 0x20018'])                                                         = <void>
+os_malloc (['10000'])                                                                     = HeapStart + 0x22360
+  brk (['HeapStart + 0x24a70'])                                                           = HeapStart + 0x24a70
```

Comment:
I'm not really sure but I think that the returned value of the `os_malloc` should have been `0x22348` which is the latest block allocated with `brk`, here:
```
os_malloc (['4000'])                                                                      = HeapStart + 0x213a8
  brk (['HeapStart + 0x22348'])                                                           = HeapStart + 0x22348
``` 

Shouldn't it expand the already existing block starting at `0x22348`? 

This is the diff between my output and ref:
```
-os_malloc (['10000'])                                                                     = HeapStart + 0x22348
-  brk (['HeapStart + 0x24a52'])                                                           = HeapStart + 0x24a52
-os_free (['HeapStart + 0x22348'])                                                         = <void>
+os_malloc (['10000'])                                                                     = HeapStart + 0x22360
+  brk (['HeapStart + 0x24a70'])                                                           = HeapStart + 0x24a70
+os_free (['HeapStart + 0x22360'])                                                         = <void>
``` 

It looks like in the ref it jumped `0x12` (sizeof of the header) from `0x22348` and because of that also `brk` requested more than needed.

Reply:
No block starts at `0x22348`. brk() returns the new program break on success (check notes from the manual[1])
> However, the actual Linux system call returns the new program break on success

The size of the structure is not `0x12 (18)`, it actually is `0x18 (24)`. 

I checked the ref file and it seems correct. I think you performed an unneeded split on the last chunk. Check the conditions about the size of the remaining size.

[1] https://man7.org/linux/man-pages/man2/brk.2.html#NOTES

Reply:
Yes, you are right, in my case the offset `30` caused the wrong behaviour. Thanks.